tutorials/ann_hnswlib.ipynb

```diff
@@ -0,0 +1,394 @@
+    "%%time\n",
+    "for uid in test_users:\n",
+    "    mf_recs.append(mf.recommend(uid, k=K))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "520da23e-79eb-4c53-9aaa-a33aaea883fc",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 312 ms, sys: 31 µs, total: 312 ms\n",
+      "Wall time: 309 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "for uid in test_users:\n",
+    "    ann_recs.append(ann.recommend(uid, k=K))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e2f4f68a-c69b-4e70-b32a-81eb78d21279",
+   "metadata": {},
+   "source": [
+    "While it took MF 5.04s to complete the task, it's only 309ms for ANN. The speed up is about 15 times. Note that our dataset contains less than 5000 items. We will see an even bigger improvement with more items and with higher dimensions of the factors."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "d9e5dde7-8012-44e7-b407-d7140047c576",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "99.75699999999999\n"
+     ]
+    }
+   ],
+   "source": [
+    "recalls = []\n",
+    "for mf_rec, ann_rec in zip(mf_recs, ann_recs):\n",
+    "    recalls.append(len(set(mf_rec) & set(ann_rec)) / len(mf_rec))\n",
+    "print(np.mean(recalls) * 100.0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bf472812-70c8-4146-b61c-b3c1f5c5b0a5",
+   "metadata": {},
+   "source": [
+    "In terms of recall, we only see a small drop of less than 1% meaning recommendations are very similar between the two. While it's almost a free lunch for this case, the numbers might differ for other cases. It's always good to make sure that ANN maintains consistent recommendations with the base model."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "daf585e5-f84b-4dc5-99f2-d6a5bd05ede4",
+   "metadata": {},
+   "source": [
+    "## Save/load for deployment"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "1fec7413-482b-417b-a12f-362a93aefb85",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'save_dir/HNSWLibANN/2023-11-09_06-15-10-660690.pkl'"
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "ann.save(\"save_dir\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "b897ead8-ad69-46d4-8903-2da65b055c54",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "loaded_ann = HNSWLibANN.load(\"save_dir/HNSWLibANN\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "74211271-2e88-4632-b18b-c87d3dfb478b",
+   "metadata": {},
+   "source": [
+    "Let's compare top-K recommendations for 5 random users between the original ANN and the loaded ANN. Of course they should be the same."
```

Comment:
- Can we double check on this? There is some randomness going on under the hood in HNSW, if we don't fix all the seeds, I am not 100% sure the results will be always the same even with a fixed index. 
- When we save and re-load HNSWLibANN we have access to all the HNSW parameters right (including the ef parameter)?

Reply:
Yes. The problem of randomness is during building the index, not save/load. So far, I don't have any problem of reproducibility with save/load the index. But it's true that we should set the `ef` param again as they suggest.

Reply:
double checked. Should be reproducible.